riptide-compression/src/main/java/org/zalando/riptide/compression/RequestCompressionPlugin.java

```diff
@@ -0,0 +1,88 @@
+package org.zalando.riptide.compression;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apiguardian.api.API;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpOutputMessage;
+import org.springframework.http.StreamingHttpOutputMessage;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestArguments.Entity;
+import org.zalando.riptide.RequestExecution;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.zip.GZIPOutputStream;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+import static org.springframework.http.HttpHeaders.CONTENT_ENCODING;
+import static org.springframework.http.HttpHeaders.TRANSFER_ENCODING;
+
+@API(status = EXPERIMENTAL)
+public final class RequestCompressionPlugin implements Plugin {
+
+    private final Compression compression;
+
+    public RequestCompressionPlugin(final Compression compression) {
+        this.compression = compression;
+    }
+
+    public RequestCompressionPlugin() {
+        this(Compression.of("gzip", GZIPOutputStream::new));
+    }
+
+    @Override
+    public RequestExecution aroundNetwork(final RequestExecution execution) {
+        return arguments -> {
+            final Entity entity = arguments.getEntity();
+
+            if (entity.isEmpty()) {
+                return execution.execute(arguments);
+            }
+
+            return execution.execute(
+                    arguments.withEntity(new CompressingEntity(compression, entity)));
+        };
+    }
+
+    @AllArgsConstructor
+    private static class CompressingEntity implements Entity {
+
+        private final Compression compression;
+        private final Entity entity;
+
+        @Override
+        public void writeTo(final HttpOutputMessage message) throws IOException {
+            update(message.getHeaders());
+
+            if (message instanceof StreamingHttpOutputMessage) {
+                final StreamingHttpOutputMessage streaming = (StreamingHttpOutputMessage) message;
```

Comment:
I'm thinking whether this could be simplified and done only once for all entities, since it's always the same pattern.